packages/socket.io-client/lib/socket.ts

```diff
@@ -440,6 +440,9 @@ export class Socket<
       packet.id = id;
     }
 
+    // check synchronously if we've missed a heartbeat, potentially causing a disconnection
+    this.io.checkHeartbeat()
```

Comment:
How about merging the `checkHeartbeat()` method with the `isTransportWritable` check below? Something like:

`packages/engine.io-client/lib/socket.ts`

```js
get writable() {
  const expired = Date.now() >= this._pingTimeoutTime
  if (expired) {
    debug("`checkHeartbeat` detected missed ping so closing connection");
    this._onClose("ping timeout");
    return false;
  }

  return this.io.engine.transport?.writable;
}
```

`packages/socket.io-client/lib/socket.ts`

```js
const isTransportWritable = this.io.engine.writable;
```

Reply:
It feels a bit strange that reading `writable` can have a side effect? Especially when it can cause events to be emitted as part of that

Maybe `writable` should return `false` when a heartbeat has been missed, but we should still have a function like `checkHeartbeat` (or maybe `disconnectIfHeartbeatMissed()`) that does the disconnection.

WDYT?

Reply:
With the `checkHeartbeat()` method, it feels like we are mixing two different layers. In my opinion, the Socket.IO client should not have to be aware of the internals of the Engine.IO client (up to a certain point, obviously).

That's why I suggested to merge the heartbeat check with the transport check below. Maybe not as a getter with side effects, but how about a method like `socket.isReady()`?

Reply:
I see. I liked the `writable` idea but I think it was the synchronous disconnection as part of reading that that felt a bit strange.

How about:
- adding `isAlive`/`isResponsive` getter which will check the last heartbeat timestamp, and if it’s `false` also schedule a disconnect on the next tick
- Maybe update `writable` too to also check the last heartbeat time and schedule a disconnection
- Instead of `this.connected` check the `isAlive` instead. (`this.connected` doesn’t work if the disconnect gets triggered on the next tick)

I think it feels better doing the disconnection on the next tick instead of sync as part of the read call, as then it’s the same pattern as if the heartbeat timer task fired anyway and there can’t be event listeners being called as part of the read

WDYT?

---

Or

We don't add `isResponsive`, and make `writable` do the check and queue the disconnect, but update the `.connected` check below to instead check `.writable`, so we always buffer if it's not writable. Not sure what impact this might have?

Reply:
Sorry for the delay.

> adding isAlive/isResponsive getter which will check the last heartbeat timestamp, and if it’s false also schedule a disconnect on the next tick

That sounds reasonable :+1:

Note: we already have a `nextTick()` method [here](https://github.com/socketio/socket.io/blob/582655f679ccc43f0a9cbef1f13ea3cde07dc2e1/packages/engine.io-client/lib/globals.ts#L1-L9).

> We don't add isResponsive, and make writable do the check and queue the disconnect, but update the .connected check below to instead check .writable, so we always buffer if it's not writable. Not sure what impact this might have?

Not sure about this one, since `writable` is an attribute of the `Transport` class, while checking the heartbeat expiry sounds like the duty of the `Socket` class.

Reply:
ok great thanks will update to that first approach then

Reply:
@darrachequesne I've pushed the changes for this